File: src/TripsModule/trips-module.js
module.exports = (function() {

var net = require('net');
var StreamPair = require('./stream-pair.js');
var KQML = require('KQML/kqml.js');

var DefaultHost = '127.0.0.1';
var DefaultPort = 6200;
var MaxPortTries = 100;

function TripsModule(argv) {
  this.argv = argv;
}
TripsModule.prototype = {
  constructor: TripsModule,

  getParameter: function(name, defalt) {
    var i = this.argv.indexOf(name);
    if (i >= 0 && i + 1 < this.argv.length) {
      return this.argv[i+1];
    } else {
      return defalt;
    }
  },

  isParameterTrue: function(name) {
    var value = this.getParameter(name, 'no');
    return /^(true|t|yes)$/i.test(value);
  },

  handleCommonParameters: function() {
    this.host = DefaultHost;
    this.port = DefaultPort;
    var value;
    if ((value = this.getParameter('-connect')) !== undefined) {
      var i;
      if (/^(true|t|yes)$/i.test(value)) {
	this.autoConnect = true;
      } else if (/^(false|nil|no)$/i.test(value)) {
	this.autoConnect = false;
      } else if ((i = value.indexOf(':')) >= 0) {
	this.host = value.slice(0,i);
	this.port = parseInt(value.slice(i+1));
      } else {
	this.host = value;
      }
    } else {
      this.autoConnect = true;
    }

    this.scanForPort = this.isParameterTrue('-scan');
    this.name = this.getParameter('-name',
      (typeof this.name == 'string') ? this.name.toLowerCase() : undefined);
    this.groupName = this.getParameter('-group',
      (typeof this.groupName == 'string') ? this.groupName.toLowerCase() : undefined);
    this.debuggingEnabled = this.isParameterTrue('-debug');
  },

  connect: function(onconnect) {
    var that = this;
    if (this.autoConnect) {
      var tries = this.scanForPort ? MaxPortTries : 1;
      // TODO detect failure to connect somehow?
      var socket = net.connect(this.port, this.host, function() {
	that.socket = KQML.wrapStream(socket);
	that.socket.on('data', function(msg) {
	  that.handleMessage(KQML.keywordify(msg));
	});
	if (onconnect !== undefined) {
	  onconnect();
	}
      });
    } else { // connect to stdin/stdout
      var pair = new StreamPair(process.stdin, process.stdout);
      this.socket = KQML.wrapStream(pair);
      this.socket.on('data', function(msg) {
	that.handleMessage(KQML.keywordify(msg));
      });
      if (onconnect !== undefined) {
	onconnect();
      }
    }
  },

  register: function() {
    if (this.name !== undefined) {
      var msg = { 0: 'register', name: this.name };
      if (this.groupName !== undefined) {
	msg.groupName = this.groupName;
      }
      this.sendMsg(msg);
    }
  },

  init: function(oninit) {
    this.handlers = {};
  //  this.pendingMessages = [];
    this.continuations = {};
    this.handleCommonParameters();
    var that = this;
    this.connect(function() {
      that.register();
      if (oninit !== undefined) {
	oninit();
      }
    });
  },

  debug: function(msg) {
    if (this.debuggingEnabled) {
      process.stderr.write(msg + "\n");
    }
  },

  addHandler: function(pattern, handler, subscribe) {
    if (subscribe === undefined || subscribe) {
      this.sendMsg({ 0: 'subscribe', content: pattern });
    }
    this.handlers[KQML.toKQML(pattern)] =
      { pattern: pattern, handler: handler };
  },

  removeHandler: function(pattern) {
    this.handlers[KQML.toKQML(pattern)] = undefined;
  },

  sendWithContinuation: function(msg, cont) {
    var replyId;
    do {
      replyId = this.name + Math.rand(1000000);
    } while (replyId in this.continuations);
    this.continuations[replyId] = cont;
    var msgWithReplyId = { replyWith: replyId };
    for (var p in msg) { msgWithReplyId[p] = msg[p] };
    this.sendMsg(msgWithReplyId);
  },

  /* TODO? not sure I *can* do this in node.js
  sendAndWait: function(msg) {
  },
  */

  sendMsg: function(msg) {
    this.socket.write(msg);
  },

  replyToMsg: function(msg, reply) {
    var modReply = { receiver: msg.sender };
    for (var p in reply) { modReply[p] = reply[p] };
    if ('replyWith' in msg) {
      modReply.inReplyTo = msg.replyWith;
    }
    this.sendMsg(modReply);
  },

  handleMessage: function(message) {
    if ('inReplyTo' in message) {
      var replyId = message.inReplyTo;
      if (replyId in this.continuations) {
	var cont = this.continuations[replyId];
	this.continuations[replyId] = undefined;
	cont(message);
      } else {
	this.sendMsg({ 0: 'error', content: ['bad-reply-id', replyId] });
      }
    } else { // not a reply, call handler
      var pat;
      for (var p in this.handlers) {
	if (KQML.patternMatches(this.handlers[p].pattern, message)) {
	  pat = p;
	  break;
	}
      }
      if (pat === undefined) {
	process.stderr.write("No handler for message: " + KQML.toKQML(message) + "\n");
      } else {
	var handler = this.handlers[p].handler;
	handler.call(this, message);
      }
    }
  },

  run: function() {
    this.sendMsg({ 0: 'tell', content: ['module-status', 'ready'] });
    // NOTE: receiving messages handled by stream ondata callbacks
  }
}

return TripsModule;

})();

File: src/TripsModule/trips_module.rb
require 'socket'
require 'KQML/kqml'

module TripsModule # two different senses of "module"
  DefaultHost = '127.0.0.1'
  DefaultPort = 6200
  MaxPortTries = 100

  include Socket::Constants

  def initialize(argv)
    @argv = argv
    @auto_connect = true
  end

  def get_parameter(param, default=nil)
    @argv.each_index { |i|
      return @argv[i+1] if (@argv[i] == param)
    }
    return default
  end

  def parameter_true?(param)
    value = get_parameter(param)
    return ((not value.nil?) and (value =~ /true|t|yes/i))
  end

  def handle_common_parameters
    @host, @port = DefaultHost, DefaultPort
    if ((value = get_parameter("-connect")) != nil)
      if (value =~ /^(?:true|t|yes)$/i)
	@auto_connect = true
      elsif (value =~ /^(?:false|nil|no)$/i)
        @auto_connect = false
      elsif (value =~ /:/)
        @host, @port = $`, $'.to_i
      else
	@host = value
      end
    else
      @auto_connect = true
    end

    @scan_for_port = parameter_true?("-scan")
    @name = get_parameter("-name", (@name==nil)? nil : @name.to_s.downcase)
    @group_name = get_parameter("-group", (@group_name==nil)? nil : @group_name.downcase)
    @debugging_enabled = parameter_true?("-debug")
  end

  def connect
    if (@auto_connect)
      tries = @scan_for_port ? MaxPortTries : 1
      begin
        @socket = Socket.new(AF_INET, SOCK_STREAM, 0)
        @socket.connect(Socket.pack_sockaddr_in(@port, @host))
	@socket.set_encoding("UTF-8")
        @kqml_in = @socket
        @kqml_out = @socket
      rescue
        tries -= 1
        if (tries > 0)
	  @port += 1
	  retry
	else
	  raise "Couldn't connect to #{@host}:\n" + $!.message
	end
      end
    else # connect to stdin/out
      @kqml_in = $stdin
      @kqml_out = $stdout
      @kqml_in.set_encoding("UTF-8")
      @kqml_out.set_encoding("UTF-8")
    end
  end

  def register
    unless (@name.nil?)
      msg = KQML[:register, :name => @name.intern]
      msg[:groupName] = @group_name unless (@group_name.nil?)
      send_msg(msg)
    end
  end

  def init
    @handlers = {}
    @pending_messages = []
    @continuations = {}
    handle_common_parameters()
    connect()
    register()
  end

  def debug(msg)
    $stderr.puts(msg) if (@debugging_enabled)
  end

  def add_handler(pattern, handler, subscribe=true)
    send_msg(KQML[:subscribe, :content => pattern]) if (subscribe)
    @handlers[pattern] = handler
  end

  def remove_handler(pattern)
    @handlers.delete(pattern)
  end

  def send_with_continuation(msg, cont)
    reply_id = nil
    begin
      reply_id = (@name + rand(1000000).to_s).intern
    end while (@continuations.key?(reply_id))
    @continuations[reply_id] = cont
    send_msg(msg + KQML[:"reply-with" => reply_id])
  end

  def send_and_wait(msg)
    reply_id = nil
    begin
      reply_id = (@name + rand(1000000).to_s).intern
    end while (@continuations.key?(reply_id))
    send_msg(msg + KQML[:"reply-with" => reply_id])
    until ((message = @kqml_in.read_kqml).nil?)
      if (message[:"in-reply-to"] == reply_id)
	return message[:content]
      else
	@pending_messages << message
      end
    end
    return nil
  end

  def send_msg(msg)
    @kqml_out.write_kqml(msg)
  end

  def handle_message(message)
    if (message.key?(:"in-reply-to"))
      reply_id = message[:"in-reply-to"]
      if (@continuations.key?(reply_id))
	cont = @continuations[reply_id]
	if (cont.nil?)
	  send_msg(KQML[:error, :content => [:"duplicate-reply", reply_id]])
	else
	  @continuations.delete(reply_id)
	  cont[message]
	end
      else
	send_msg(KQML[:error, :content => [:"bad-reply-id", reply_id]])
      end
    else # not a reply, call handler
      pat = @handlers.keys.find { |pattern| pattern === message }
      if (pat.nil?)
	$stderr.puts "No handler for message: " + message.to_kqml_string
      else
	@handlers[pat].call(message)
      end
    end
  end

  def get_next_message
    @pending_messages.shift || @kqml_in.read_kqml
  end

  def run
    send_msg(KQML[:tell, :content => [:"module-status", :ready]])
    while ((message = get_next_message) != nil)
      begin
        handle_message(message)
      rescue
        send_msg(KQML[:error, :content => [$!.to_s, $!.backtrace.join("\n")]] + KQML.in_reply_to(message))
      end
    end
  end
end

